core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java

```diff
@@ -0,0 +1,1793 @@
+      this.name = cacheName;
+   }
+
+   @Inject
+   public void injectDependencies(ComponentRegistry componentRegistry,
+                                  Configuration configuration,
+                                  EmbeddedCacheManager cacheManager,
+                                  DataContainer dataContainer,
+                                  CacheNotifier cacheNotifier,
+                                  TimeService timeService,
+                                  InternalEntryFactory entryFactory) {
+      this.componentRegistry = componentRegistry;
+      this.configuration = configuration;
+      this.cacheManager = cacheManager;
+      this.dataContainer = dataContainer;
+      this.cacheNotifier = cacheNotifier;
+      this.timeService = timeService;
+      this.entryFactory = entryFactory;
+   }
+
+   @Override
+   @ManagedOperation(
+         description = "Starts the cache.",
+         displayName = "Starts cache."
+   )
+   public void start() {
+      this.defaultMetadata = new EmbeddedMetadata.Builder()
+            .lifespan(configuration.expiration().lifespan())
+            .maxIdle(configuration.expiration().maxIdle()).build();
+      this.keyEquivalence = configuration.dataContainer().keyEquivalence();
+      this.valueEquivalence = configuration.dataContainer().valueEquivalence();
+      this.entryEquivalence = new EntryEquivalence(keyEquivalence, valueEquivalence);
+      componentRegistry.start();
+   }
+
+   @Override
+   @ManagedOperation(
+         description = "Stops the cache.",
+         displayName = "Stops cache."
+   )
+   public void stop() {
+      dataContainer = null;
+      componentRegistry.stop();
+   }
+
+
+   @Override
+   public void putForExternalRead(K key, V value) {
+      putForExternalReadInternal(key, value, defaultMetadata);
+   }
+
+   @Override
+   public void putForExternalRead(K key, V value, long lifespan, TimeUnit unit) {
+      Metadata metadata = createMetadata(lifespan, unit);
+      putForExternalReadInternal(key, value, metadata);
+   }
+
+   @Override
+   public void putForExternalRead(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdle, TimeUnit maxIdleUnit) {
+      Metadata metadata = createMetadata(lifespan, lifespanUnit, maxIdle, maxIdleUnit);
+      putForExternalReadInternal(key, value, metadata);
+   }
+
+   @Override
+   public void putForExternalRead(K key, V value, Metadata metadata) {
+      putForExternalReadInternal(key, value, applyDefaultMetadata(metadata));
+   }
+
+   private void putForExternalReadInternal(K key, V value, Metadata metadata) {
+      Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED);
+      Objects.requireNonNull(value, NULL_VALUES_NOT_SUPPORTED);
+      ByRef.Boolean isCreatedRef = new ByRef.Boolean(false);
+      putForExternalReadInternal(key, value, metadata, isCreatedRef);
+   }
+
+   protected void putForExternalReadInternal(K key, V value, Metadata metadata, ByRef.Boolean isCreatedRef) {
```

Comment:
I think the two `putForExternalReadInternal` methods could be merged, or at least one of them should have its name changed.